src/common/KokkosKernels_IOUtils.hpp

```diff
@@ -413,61 +412,113 @@ crsMat_t kk_generate_triangular_sparse_matrix(
+  values_view_t &values_view) {
+
+  // typedef typename row_map_view_t::non_const_value_type size_type;
+  // typedef typename cols_view_t::non_const_value_type lno_t;
   typedef typename values_view_t::non_const_value_type scalar_t;
   lno_t *adj;
-  size_type *xadj;//, nnzA;
+  size_type *xadj;
   scalar_t *values;
 
   kk_sparseMatrix_generate<scalar_t, lno_t, size_type>(
       nrows, ncols, nnz, row_size_variance,  bandwidth,
-      values, xadj, adj);
-
-  row_map_view_t rowmap_view("rowmap_view", nrows+1);
-  cols_view_t columns_view("colsmap_view", nnz);
-  values_view_t values_view("values_view", nnz);
+      values, xadj, adj, block_dim * block_dim);
 
+  const lno_t nvals = nnz * block_dim * block_dim;
+  rowmap_view = row_map_view_t("rowmap_view", nrows + 1);
+  columns_view = cols_view_t("colsmap_view", nnz);
+  values_view = values_view_t("values_view", nvals);
   {
     typename row_map_view_t::HostMirror hr = Kokkos::create_mirror_view (rowmap_view);
     typename cols_view_t::HostMirror hc = Kokkos::create_mirror_view (columns_view);
     typename values_view_t::HostMirror hv = Kokkos::create_mirror_view (values_view);
 
-    for (lno_t i = 0; i <= nrows; ++i){
+    Kokkos::parallel_for(nrows + 1, KOKKOS_LAMBDA(const lno_t i) {
       hr(i) = xadj[i];
-    }
-
-    for (size_type i = 0; i < nnz; ++i){
+    });
+    Kokkos::parallel_for(nnz, KOKKOS_LAMBDA(const size_type i) {
       hc(i) = adj[i];
+    });
+    Kokkos::parallel_for(nvals, KOKKOS_LAMBDA(const size_type i) {
       hv(i) = values[i];
-    }
+    });
     Kokkos::deep_copy (rowmap_view , hr);
     Kokkos::deep_copy (columns_view , hc);
     Kokkos::deep_copy (values_view , hv);
+    Kokkos::fence();
   }
+  delete [] xadj;
+  delete [] adj;
+  delete [] values;
+}
 
-  graph_t static_graph (columns_view, rowmap_view);
+template <typename crsMat_t>
+crsMat_t kk_generate_sparse_matrix(
+    typename crsMat_t::const_ordinal_type nrows,
+    typename crsMat_t::const_ordinal_type ncols,
+    typename crsMat_t::non_const_size_type &nnz,
+    typename crsMat_t::const_ordinal_type row_size_variance,
+    typename crsMat_t::const_ordinal_type bandwidth){
+
+  typedef typename crsMat_t::StaticCrsGraphType graph_t;
+  typedef typename graph_t::row_map_type::non_const_type row_map_view_t;
+  typedef typename graph_t::entries_type::non_const_type cols_view_t;
+  typedef typename crsMat_t::values_type::non_const_type values_view_t;
+
+  row_map_view_t rowmap_view;
+  cols_view_t columns_view;
+  values_view_t values_view;
+
+  // Note: block_dim=1 generates CRS content
+  generate_bsr_content(1, nrows, ncols, nnz, row_size_variance,
+    bandwidth, rowmap_view, columns_view, values_view);
+
+  typename crsMat_t::StaticCrsGraphType static_graph(columns_view, rowmap_view);
   crsMat_t crsmat("CrsMatrix", ncols, values_view, static_graph);
-  delete [] xadj; delete [] adj; delete [] values;
   return crsmat;
 }
 
+template <typename bsrMat_t>
+bsrMat_t kk_generate_sparse_matrix(
```

Comment:
Naive question: could you fill first a CrsMatrix and then use the BsrMatrix constructor that takes a CrsMatrix and a block dimension as input parameters?

Reply:
Yes, I reverted to this simpler approach